//
// Copyright 2019 The Project Oak Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

syntax = "proto3";

package oak;

import "google/protobuf/any.proto";
import "third_party/google/rpc/status.proto";

// Protocol buffer encoding to hold additional metadata that accompanies a gRPC
// message exchange.  This is normally encoded using various HTTP2 features
// (https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) but to avoid
// the need for a full HTTP2 implementation inside each Oak Node, we define a
// simplified envelope format that includes the relevant details.

message GrpcRequest {
  int32 stream_id = 1;
  string method_name = 2;
  google.protobuf.Any req_msg = 3;
  bool last = 4;
}

message GrpcResponse {
  int32 stream_id = 1;
  google.protobuf.Any rsp_msg = 2;
  google.rpc.Status status = 3;
  bool last = 4;
}
